nifi-nar-bundles/nifi-hashicorp-vault-bundle/nifi-hashicorp-vault-client-service/src/main/java/org/apache/nifi/vault/hashicorp/StandardHashiCorpVaultClientService.java

```diff
@@ -0,0 +1,209 @@
+                    .build());
+        } catch (final Exception e) {
+            verificationLogger.error("Failed to configure HashiCorp Vault Client", e);
+
+            results.add(new ConfigVerificationResult.Builder()
+                    .outcome(ConfigVerificationResult.Outcome.FAILED)
+                    .verificationStepName("Configure HashiCorp Vault Client")
+                    .explanation("Failed to configure HashiCorp Vault Client: " + e.getMessage())
+                    .build());
+        }
+        if (service != null) {
+            try {
+                service.testConnection();
+                results.add(new ConfigVerificationResult.Builder()
+                        .outcome(ConfigVerificationResult.Outcome.SUCCESSFUL)
+                        .verificationStepName("Connect to HashiCorp Vault Server")
+                        .explanation("Successfully connected to HashiCorp Vault Server")
+                        .build());
+            } catch (final Exception e) {
+                verificationLogger.error("Failed to connect to HashiCorp Vault Server", e);
+
+                results.add(new ConfigVerificationResult.Builder()
+                        .outcome(ConfigVerificationResult.Outcome.FAILED)
+                        .verificationStepName("Connect to HashiCorp Vault Server")
+                        .explanation("Failed to connect to HashiCorp Vault Server: " + e.getMessage())
+                        .build());
+            }
+        }
+
+        return results;
+    }
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) throws InitializationException {
+        try {
+            communicationService = createCommunicationService(context);
+        } catch (final Exception e) {
+            throw new InitializationException("Failed to initialize HashiCorp Vault client", e);
+        }
+    }
+
+    @OnDisabled
+    public void onDisabled() {
+        communicationService = null;
+    }
+
+    @Override
+    public HashiCorpVaultCommunicationService getHashiCorpVaultCommunicationService() {
+        return communicationService;
+    }
+
+    private HashiCorpVaultCommunicationService createCommunicationService(final ConfigurationContext context) throws IOException {
+        final List<PropertySource<?>> propertySources = new ArrayList<>();
+
+        final String configurationStrategy = context.getProperty(CONFIGURATION_STRATEGY).getValue();
+        if (DIRECT_PROPERTIES.getValue().equals(configurationStrategy)) {
+            final PropertySource<?> configurationPropertySource = new DirectPropertySource("Direct Properties", context);
+            propertySources.add(configurationPropertySource);
+        } else {
+            final String propertiesFiles = context.getProperty(VAULT_PROPERTIES_FILES).getValue();
+            for (final String propertiesFile : propertiesFiles.split(COMMA_SEPARATOR)) {
```

Comment:
This approach can be changed to use the `asResources()` and Resource References feature of Property Descriptors.